src/xbase/layout/mos/placement/compact.py

```diff
@@ -254,32 +293,73 @@ def _place_rows(tr_manager: TrackManager, tech_cls: MOSTech, pspecs_list: Sequen
-                                 ytop_conn=max(conn_y_bnds_bot[1], conn_y_bnds_top[1]))
-            if row_info.row_type.is_substrate:
-                top_wg.align_wires(hm_layer, tr_manager, ytop_prev, ycur + blk_h, top_pcons=None)
-
-        # get ytop_conn
-        bnd_table = top_wg.get_placement_bounds(hm_layer, grid)
-        conn_y_bnds_top[0] = _update_y_conn(grid, row_info, conn_layer, ycur, bnd_table,
-                                            conn_y_bnds_top[0], False)
-        conn_y_bnds_top[1] = _update_y_conn(grid, row_info, conn_layer, ycur, bnd_table,
-                                            conn_y_bnds_top[1], True)
+
+        is_top_row = (idx == num_rows - 1)
+        if row_specs.double_gate:
+            # place middle wires
+            conn_y_bnds_mid = [COORD_MAX, COORD_MIN]
+            pcons = {}
+            for wtype, conn_y in mid_conn_y_table.items():
+                pcons[wtype.name] = ycur + conn_y[0]
+                if wtype.is_physical:
+                    conn_y_bnds_mid[0] = min(conn_y_bnds_mid[0], ycur + conn_y[0])
+                    conn_y_bnds_mid[1] = max(conn_y_bnds_mid[1], ycur + conn_y[1])
+
+            if mid_wg:
+                mid_wg.place_compact(hm_layer, tr_manager,
+                                     pcons=PlaceFun(conn_layer, grid, pcons, RoundMode.GREATER_EQ),
+                                     prev_wg=prev_wg, #top_mirror=top_mirror and is_top_row,)
+                                     ytop_conn=max(conn_y_bnds_bot[1], conn_y_bnds_mid[1]))
+                prev_wg = mid_wg
+            
+            # place top wires above the middle wires
+            conn_y_bnds_top = [COORD_MAX, COORD_MIN]
+            pcons = {}
+            for wtype, conn_y in top_conn_y_table.items():
+                pcons[wtype.name] = ycur + conn_y[0]
+                if wtype.is_physical:
+                    conn_y_bnds_top[0] = min(conn_y_bnds_top[0], ycur + conn_y[0])
+                    conn_y_bnds_top[1] = max(conn_y_bnds_top[1], ycur + conn_y[1])
+
+            if top_wg:
+                top_wg.place_compact(hm_layer, tr_manager,
+                                     pcons=PlaceFun(conn_layer, grid, pcons, RoundMode.GREATER_EQ),
+                                     prev_wg=prev_wg, #top_mirror=top_mirror and is_top_row,
+                                     ytop_conn=max(conn_y_bnds_bot[1], conn_y_bnds_top[1]))
+                prev_wg = top_wg
+            
+            bnd_table = top_wg.get_placement_bounds(hm_layer, grid)
+            conn_y_bnds_top[0] = _update_y_conn(grid, row_info, conn_layer, ycur, bnd_table,
+                                                conn_y_bnds_top[0], False)
+            conn_y_bnds_top[1] = _update_y_conn(grid, row_info, conn_layer, ycur, bnd_table,
+                                                conn_y_bnds_top[1], True)
+        else:
+            # place top wires
+            
```

Comment:
remove space - not critical, but might as well do it since you are making the other changes